src/NHibernate/Engine/QueryCacheBatchQueue.cs

```diff
@@ -0,0 +1,271 @@
+		/// Once this method is called the uninitialized collection keys for a given role will be cleared in order to prevent
+		/// double checking the same keys.
+		/// </summary>
+		/// <param name="collectionPersister">The persister for the collection role.</param>
+		/// <param name="key">A key that must be included in the batch fetch.</param>
+		/// <param name="collectionEntries">An array that will be filled with collection entries if set.</param>
+		/// <returns>
+		/// An array of collection keys that can be empty if the key was already checked or <see langword="null" />
+		/// if the key is not present in the cached query.
+		/// </returns>
+		internal object[] GetCollectionBatch(ICollectionPersister collectionPersister, object key, out CollectionEntry[] collectionEntries)
+		{
+			if (!_queryCollectionKeys.TryGetValue(collectionPersister.Role, out var keys))
+			{
+				collectionEntries = null;
+				return null; // The collection was not present in the cached query
+			}
+
+			var collectionKey = new CollectionKey(collectionPersister, key);
+			if (_queryCheckedCollectionKeys.TryGetValue(collectionPersister.Role, out var checkedKeys) &&
+				checkedKeys.Contains(collectionKey))
+			{
+				collectionEntries = null;
+				return Array.Empty<object>();
+			}
+
+			if (!keys.TryGetValue(collectionKey, out var collectionEntry) || collectionEntry == null)
+			{
+				collectionEntries = null;
+				return null; // The collection was not present in the cached query
+			}
+
+			if (checkedKeys == null)
+			{
+				checkedKeys = new HashSet<CollectionKey>();
+				_queryCheckedCollectionKeys.Add(collectionPersister.Role, checkedKeys);
+			}
+
+			if (!_queryCheckedCollectionEntries.TryGetValue(collectionPersister.Role, out var checkedEntries))
+			{
+				checkedEntries = new HashSet<CollectionEntry>();
+				_queryCheckedCollectionEntries.Add(collectionPersister.Role, checkedEntries);
+			}
+
+			var result = new object[keys.Count];
+			collectionEntries = new CollectionEntry[result.Length];
+			var i = 0;
+			result[i++] = key;
+
+			foreach (var pair in keys)
+			{
+				if (pair.Value == null || _persistenceContext.GetCollection(pair.Key)?.WasInitialized != false)
+				{
+					continue; // The collection was not registered or is already initialized
+				}
+
+				if (collectionPersister.KeyType.IsEqual(key, pair.Value.LoadedKey, collectionPersister.Factory))
+				{
+					collectionEntries[0] = pair.Value;
+					checkedKeys.Add(pair.Key);
+					checkedEntries.Add(pair.Value);
+					continue;
+				}
+
+				collectionEntries[i] = pair.Value;
+				result[i++] = pair.Value.LoadedKey;
+				checkedKeys.Add(pair.Key);
+				checkedEntries.Add(pair.Value);
+			}
+
+			keys.Clear();
+
+			return result;
+		}
+
+		/// <summary>
+		/// Adds the entity to the batch.
+		/// </summary>
+		/// <param name="key">The entity key.</param>
+		internal void AddEntityKey(EntityKey key)
+		{
+			if (!key.IsBatchLoadable)
```

Comment:
This condition is indeed `batchSize <= 1`. In the context of caching or retrieving from cache a query's results, I do not think `batchSize` is relevant. `batchSize` is relevant to lazy loads, it decides whether to pro-actively load more data than directly asked by the current operation or not.
For a cacheable query, there is no reason to prevent a multiple get of its results, they will all be attempted anyway.
But well, removing the condition here will not change anything currently, since anyway the call to `AddEntityKey` is already ruled-out by the same condition in `ManyToOneType`.

It seems to me all the already existing checks on `batchSize` should be disabled/avoided by this PR when a query cache batch queue is enabled, and no new ones should be added. But maybe doing this may add to much complexity.

Reply:
I've modified the code to allow batching regardless of `batchSize` and I don't think that it added too much complexity.